### PR TITLE
urlapi: return CURLUE_BAD_HOSTNAME if puny2idn encoding fails

### PR DESCRIPTION
**Pull request by bagder**
And document it

### CODE REVIEW

**Comment by bagder** [comment]
Alternatives:
1. have the flag be a "best effort" and just return it as-is if it fails (but that then doesn't tell the user if it worked)
2. introduce a new error code to make the IDN conversion failure more explicit
3. keep this PR

**Comment by bagder** [comment]
@jacobmealey thoughts?

**Comment by jacobmealey** [comment]
it looks like `idn2_to_unicode_8z8z(puny, &enc, 0)` is never throwing an error when passed some malformed punycode. It _should_ return [IDN2_ENCODING_ERROR](https://www.gnu.org/software/libidn/libidn2/reference/libidn2-idn2.h.html#IDN2-ENCODING-ERROR:CAPS) but its returning 0 (IDN2_OK) and an empty string. I've only done preliminary testing but that's what I've noticed. 

note that I'm using `libidn2.so.0.3.8`

**Comment by bagder** [comment]
we could of course also remove the check for the `xn--` prefix and then trow a string at it that does not start with `xn--`...

**Comment by bagder** [comment]
> it may be good to have curl_url_get return the same error for Curl_idn_decode

Ack. I'll make sure that it returns CURLUE_BAD_HOSTNAME for bad host names and CURLUE_OUT_OF_MEMORY only when it actually runs out of memory.

File: lib/idn.c [hunk 1]
```diff
@@ -157,52 +173,68 @@ static char *idn_decode(const char *input)
...
+  if(!result)
+    *output = d;
+  return result;
 }
 
-char *Curl_idn_encode(const char *puny)
+CURLcode Curl_idn_encode(const char *puny, char **output)
 {
-  char *d = idn_encode(puny);
+  char *d;
```
Comment (jacobmealey) [inline lib/idn.c:223]: This either needs to be declared as NULL or the if statement on line 226 should be checking the result. Currently on a failed idn conversion this causes a `free(): invalid pointer` error on my machine. Changing the if statement or declaring d as NULL fixed it for me

Comment (bagder) [reply lib/idn.c]: :+1:

**Pull request closed without merging**